libp2p/crypto/crypto.nim

```diff
@@ -325,50 +347,77 @@ proc init*(key: var PrivateKey, data: string): bool =
   ## hexadecimal string representation.
   ##
   ## Returns ``true`` on success.
-  result = key.init(fromHex(data))
+  try:
+    key.init(fromHex(data))
+  except ValueError:
+    false
 
 proc init*(key: var PublicKey, data: string): bool =
   ## Initialize public key ``key`` from libp2p's protobuf serialized
   ## hexadecimal string representation.
   ##
   ## Returns ``true`` on success.
-  result = key.init(fromHex(data))
+  try:
+    key.init(fromHex(data))
+  except ValueError:
+    false
 
 proc init*(sig: var Signature, data: string): bool =
   ## Initialize signature ``sig`` from serialized hexadecimal string
   ## representation.
   ##
   ## Returns ``true`` on success.
-  result = sig.init(fromHex(data))
+  try:
+    sig.init(fromHex(data))
+  except ValueError:
+    false
 
-proc init*(t: typedesc[PrivateKey], data: openarray[byte]): PrivateKey =
+proc init*(t: typedesc[PrivateKey], data: openarray[byte]): CryptoResult[PrivateKey] =
   ## Create new private key from libp2p's protobuf serialized binary form.
-  if not result.init(data):
-    raise newException(P2pKeyError, "Incorrect binary form")
+  var res: t
+  if not res.init(data):
+    err(KeyError)
+  else:
+    ok(res)
 
-proc init*(t: typedesc[PublicKey], data: openarray[byte]): PublicKey =
+proc init*(t: typedesc[PublicKey], data: openarray[byte]): CryptoResult[PublicKey] =
   ## Create new public key from libp2p's protobuf serialized binary form.
-  if not result.init(data):
-    raise newException(P2pKeyError, "Incorrect binary form")
+  var res: t
+  if not res.init(data):
+    err(KeyError)
+  else:
+    ok(res)
 
-proc init*(t: typedesc[Signature], data: openarray[byte]): Signature =
+proc init*(t: typedesc[Signature], data: openarray[byte]): CryptoResult[Signature] =
   ## Create new public key from libp2p's protobuf serialized binary form.
-  if not result.init(data):
-    raise newException(P2pSigError, "Incorrect binary form")
+  var res: t
+  if not res.init(data):
+    err(SigError)
+  else:
+    ok(res)
 
-proc init*(t: typedesc[PrivateKey], data: string): PrivateKey =
+proc init*(t: typedesc[PrivateKey], data: string): CryptoResult[PrivateKey] =
   ## Create new private key from libp2p's protobuf serialized hexadecimal string
   ## form.
-  result = t.init(fromHex(data))
+  try:
+    t.init(fromHex(data))
+  except ValueError:
+    err(KeyError)
 
-proc init*(t: typedesc[PublicKey], data: string): PublicKey =
+proc init*(t: typedesc[PublicKey], data: string): CryptoResult[PublicKey] =
   ## Create new public key from libp2p's protobuf serialized hexadecimal string
   ## form.
-  result = t.init(fromHex(data))
+  try:
+    t.init(fromHex(data))
+  except ValueError:
+    err(KeyError)
 
-proc init*(t: typedesc[Signature], data: string): Signature =
+proc init*(t: typedesc[Signature], data: string): CryptoResult[Signature] =
```

Comment:
at some point, it'd be nice if these were called `fromHex` because we also have `DER`, `baseXX` and binary floating around and it's not at at all clear which one the string is